test/json_unittest.cc

```diff
@@ -0,0 +1,339 @@
+  });
+}
+
+TEST(EdgeTest, StringWithUnicodeEscape) {
+  GuardJsonException([](){
+    std::unique_ptr<json::Value> v = json::Parser::FromString("\"foo\\u000abar\"");
+    //EXPECT_EQ(v->As<json::String>()->value(), "foo\nbar");
+    // Output should be escaped as that is more canonical.
+    EXPECT_TO_STRING(v, "\"foo\\nbar\"");
+  });
+}
+
+TEST(EdgeTest, StringWithUTF8) {
+  GuardJsonException([](){
+    // This is Korean for "Hello World!".
+    std::unique_ptr<json::Value> v = json::Parser::FromString("\"안녕 세상아!\"");
+    EXPECT_EQ(v->As<json::String>()->value(), "안녕 세상아!");
+    EXPECT_TO_STRING(v, "\"안녕 세상아!\"");
+  });
+}
+
+
+// Number edge cases
+
+TEST(EdgeTest, PositiveNumbers) {
+  GuardJsonException([](){
+    std::unique_ptr<json::Value> v = json::Parser::FromString(
+      "[\n"
+      "  0, 0e+0, 0e-0, 0e0, 0E+0, 0E-0, 0E0,\n"
+      "  0.0, 0.0e+0, 0.0e-0, 0.0e0, 0.0E+0, 0.0E-0, 0.0E0,\n"
+      "  10, 10e+0, 10e-0, 10e0, 10E+0, 10E-0, 10E0,\n"
+      "  10.0, 10.0e+0, 10.0e-0, 10.0e0, 10.0E+0, 10.0E-0, 10.0E0,\n"
+      "  123, 123.456, 789, 1.05, 1.999e-99\n"
+      "]"
+    );
+    std::vector<double> numbers;
+    for (const auto& number : *v->As<json::Array>()) {
+      numbers.push_back(number->As<json::Number>()->value());
+    }
+    EXPECT_EQ(numbers, std::vector<double>({
+      0, 0e+0, 0e-0, 0e0, 0E+0, 0E-0, 0E0,
+      0.0, 0.0e+0, 0.0e-0, 0.0e0, 0.0E+0, 0.0E-0, 0.0E0,
+      10, 10e+0, 10e-0, 10e0, 10E+0, 10E-0, 10E0,
+      10.0, 10.0e+0, 10.0e-0, 10.0e0, 10.0E+0, 10.0E-0, 10.0E0,
+      123, 123.456, 789, 1.05, 1.999e-99,
+    }));
+    EXPECT_TO_STRING(v,
+      "["
+        "0.0,0.0,0.0,0.0,0.0,0.0,0.0,"
+        "0.0,0.0,0.0,0.0,0.0,0.0,0.0,"
+        "10.0,10.0,10.0,10.0,10.0,10.0,10.0,"
+        "10.0,10.0,10.0,10.0,10.0,10.0,10.0,"
+        "123.0,123.45600000000000307,789.0,"
+        "1.0500000000000000444,1.9989999999999998999e-99"
+      "]"
+    );
+  });
+}
+
+TEST(EdgeTest, NegativeNumbers) {
+  GuardJsonException([](){
+    std::unique_ptr<json::Value> v = json::Parser::FromString(
+      "[\n"
+      "  -0, -0e+0, -0e-0, -0e0, -0E+0, -0E-0, -0E0,\n"
+      "  -0.0, -0.0e+0, -0.0e-0, -0.0e0, -0.0E+0, -0.0E-0, -0.0E0,\n"
+      "  -10, -10e+0, -10e-0, -10e0, -10E+0, -10E-0, -10E0,\n"
+      "  -10.0, -10.0e+0, -10.0e-0, -10.0e0, -10.0E+0, -10.0E-0, -10.0E0,\n"
+      "  -123, -123.456, -789, -1.05, -1.999e-99\n"
+      "]"
+    );
+    std::vector<double> numbers;
+    for (const auto& number : *v->As<json::Array>()) {
+      numbers.push_back(number->As<json::Number>()->value());
+    }
+    EXPECT_EQ(numbers, std::vector<double>({
+      -0, -0e+0, -0e-0, -0e0, -0E+0, -0E-0, -0E0,
+      -0.0, -0.0e+0, -0.0e-0, -0.0e0, -0.0E+0, -0.0E-0, -0.0E0,
+      -10, -10e+0, -10e-0, -10e0, -10E+0, -10E-0, -10E0,
+      -10.0, -10.0e+0, -10.0e-0, -10.0e0, -10.0E+0, -10.0E-0, -10.0E0,
+      -123, -123.456, -789, -1.05, -1.999e-99,
+    }));
+    // TODO(igorp): This first one looks spurious.
```

Comment:
@igorpeshansky

Reply:
Sorry, which first one? Do you mean `-0` transforming to `0.0`? That's because with `int`, there is no separate representation for `-0`, so it gets converted to `static_cast<double>(0)`. Nothing to do here.

Reply:
That doesn't explain json::Parser::FromString("-0")->ToString() == "0"

Reply:
Hmm, true. Ok, warrants further investigation. Let's remove the username from the TODO, though, and add what you just said...